src/dvc_data/hashfile/gc.py

```diff
@@ -23,18 +23,22 @@ def _is_dir_hash(_hash):
         return _hash.endswith(HASH_DIR_SUFFIX)
 
     removed = False
-    # hashes must be sorted to ensure we always remove .dir files first
```

Comment:
Just doubt I have, why `.dir` files have to be removed first?

Reply:
We consider `.dir` files (and tracked objects inside) to be one single unit. We use this for optimization (assumptions such as if `.dir` file exists in the remote, all the directory contents also exist, etc). 

https://github.com/iterative/dvc-data/blob/main/src/dvc_data/hashfile/status.py#L63-L64

We also try to upload all directory contents, and then only if they succeed, we upload `.dir` files.

So ideally here, we should try to delete `.dir` file first, and then all of it's contents in bulk and so on for all objects. Then again do bulk for `HashFile` objects.

Reply:
> So ideally here, we should try to delete .dir file first, and then all of it's contents in bulk and so on for all objects. Then again do bulk for HashFile objects.

Should we open a separate issue for that?

This P.R. just keeps the previous behavior but benefits from filesystems that implement bulk delete like `gcfs` and `s3fs`

Reply:
@daavoo You probably don't want to remove that comment completely.

Reply:
Should it be better a `# TODO:` comment to implement what @skshetry described above?

Reply:
@daavoo, let's just leave it out. It's harder to do it in practice (Eg: what if there are overlaps between different `.dir` files, etc.).

It is unsafe, but `gc` always is.

Reply:
As long as we remove .dir files before all other files first then gc should be safe. The optimization isn't affected by overlaps between dirs. The only thing we assume is that when checking remote status is that if a .dir file exists on the remote, then all files inside that .dir exist. It's more unsafe for us to actually attempt to remove .dir+directory contents, then another .dir+directory contents, and so on, because then we do have to get into figuring out directory overlaps.

The existing behavior (and the current state of the PR) should be fine for now.